hail/python/test/hailtop/batch/test_batch.py

```diff
@@ -88,6 +88,26 @@ def test_single_job(self):
 
             assert self.read(output_file.name) == msg
 
+    def test_single_job_with_shell(self):
+        with tempfile.NamedTemporaryFile('w') as output_file:
+            msg = 'hello world'
+
+            b = self.batch()
+            j = b.new_job(shell='/bin/sh')
+            j.command(f'echo "{msg}" > {j.ofile}')
+
+            b.write_output(j.ofile, output_file.name)
+            b.run()
+
+            assert self.read(output_file.name) == msg
+
+    def test_single_job_with_nonsense_shell(self):
+        b = self.batch()
+        j = b.new_job(shell='/bin/ajdsfoijasidojf')
+        j.image('google/cloud-sdk:237.0.0-alpine')
+        j.command(f'echo "hello"')
+        self.assertRaises(Exception, b.run)
```

Comment:
This code seems fine, but I'm baffled by something.  Was this test passing when you had `b.run()`?  I explicitly checked and thought so.  If `b.run()` was throwing, that would make the test fail. So I assume `b.run()` wasn't throwing.  But now this is passing, and I just verified locally `b.run()` is throwing in this case.  So why wasn't the test case failing before?  I'm afraid it isn't being run or checked correctly.  Do you know what's happening here?

Reply:
It was the missing `j.image(..)`! Without image specified we override nothing, because the command is directly invoked in the system shell (well, as elaborated below in response to the question on the use of j.image, I am not overriding sp.check_call's shell, which defaults to /bin/sh on POSIX system). I thought we wouldn't want to change this behavior, because prior to this PR we were not specifying a shell executable for jobs that did not specify`image`.

Reply:
OK!  I'm glad I asked.  I think you should change the other case, too.  I think this is a mistake.  The goal is to run exactly the same command but locally instead of inside a container.  This is bad because you get totally different shell behavior inside and outside a container, and shell is ignored depending on the setting of an independent value (input).  Let's get two tests, image or not.

batch has many fewer users and is much more experimental than the Hail interface.  I don't view the interface is immutable.  @jigold are you OK with this?

Reply:
I'm okay with changes to the interface as long as you post the changes to the Zulip forum and a mailing list that we should create (does not exist yet).

Reply:
And it should go in the change log.  Are CHANGELOG commit messages the right thing for the Batch change log, too?

Reply:
Agreed. The change log behaves the exact same as for Hail Query. Whomever makes the release is supposed to sort out which change logs go to the correct location.